src/Specular/Ref.purs

```diff
@@ -0,0 +1,119 @@
+module Specular.Ref
+  ( Ref(..)
+  , newRef
+  , newRefWithEvent
+  , refValue
+  , refUpdate
+  , refUpdateConst
+  , focusRef
+  , pureFocusRef
+  , previewRef
+  , updateRef
+  , readRef
+
+  , constRef
+
+  , wrapViewWidget
+
+  , Lens
+  , Prism
+  ) where
+
+import Prelude
+
+import Control.Apply (lift2)
+import Control.Monad.Cleanup (runCleanupT)
+import Data.Functor.Contravariant (cmap, (>$<))
+import Data.Functor.Invariant (class Invariant)
+import Data.Maybe (Maybe(..))
+import Data.Tuple (fst)
+import Effect (Effect)
+import Effect.Class (class MonadEffect)
+import Specular.Callback (Callback, attachEvent, contramapCallbackDyn, newCallback, nullCallback, triggerCallback)
+import Specular.Dom.Widget (class MonadWidget)
+import Specular.FRP (class MonadFRP, Dynamic, Event, WeakDynamic, current, foldDyn, leftmost, pull, readBehavior, weaken)
+
+
+data Ref a = Ref (Dynamic a) (Callback (a -> a))
+
+instance invariantRef :: Invariant Ref where
+  imap f g (Ref value update) = Ref (f <$> value) (cmap (\h -> g <<< h <<< f) update)
+
+newRef :: forall m a. MonadEffect m => a -> m (Ref a)
+newRef initial = do
+
+  -- Note: It's morally correct here to ignore cleanup action,
+  -- because we're not subscribing to any outside events.
+  --
+  -- It will interact badly with the leak checker, but there's no easy
```

Comment:
There is a solution now: replace `newEvent+foldDyn` with `newDynamic` (it returns a `modify` function).